pkg/transform/gtmss_preview.go

```diff
@@ -0,0 +1,115 @@
+package transform
+
+import (
+	"errors"
+
+	"github.com/snowplow/snowbridge/config"
+	"github.com/snowplow/snowbridge/pkg/models"
+
+	"github.com/snowplow/snowplow-golang-analytics-sdk/analytics"
+)
+
+// The gtmssPreviewAdapter implements the Pluggable interface
+type gtmssPreviewAdapter func(i interface{}) (interface{}, error)
+
+// ProvideDefault implements the ComponentConfigurable interface
+func (f gtmssPreviewAdapter) ProvideDefault() (interface{}, error) {
+	return nil, nil
+}
+
+// Create implements the ComponentCreator interface.
+func (f gtmssPreviewAdapter) Create(i interface{}) (interface{}, error) {
+	return f(i)
+}
+
+// gtmssPreviewAdapterGenerator returns a gtmssPreviewAdapter
+func gtmssPreviewAdapterGenerator(f func() (TransformationFunction, error)) gtmssPreviewAdapter {
+	return func(i interface{}) (interface{}, error) {
+		if i != nil {
+			return nil, errors.New("unexpected configuration input for gtmssPreview transformation")
+		}
+
+		return f()
+	}
+}
+
+// gtmssPreviewConfigFunction returns a transformation function
+func gtmssPreviewConfigFunction() (TransformationFunction, error) {
+	ctx := "contexts_com_google_tag-manager_server-side_preview_mode_1"
+	property := "x-gtm-server-preview"
+	header := "x-gtm-server-preview"
+	return gtmssPreviewTransformation(ctx, property, header), nil
+}
+
+// GTMSSPreviewConfigPair is the configuration pair for the gtmss preview transformation
+var GTMSSPreviewConfigPair = config.ConfigurationPair{
+	Name:   "gtmssPreview",
+	Handle: gtmssPreviewAdapterGenerator(gtmssPreviewConfigFunction),
+}
+
+// gtmssPreviewTransformation returns a transformation function
+func gtmssPreviewTransformation(ctx, property, headerKey string) TransformationFunction {
+	return func(message *models.Message, interState interface{}) (*models.Message, *models.Message, *models.Message, interface{}) {
+		parsedEvent, err := IntermediateAsSpEnrichedParsed(interState, message)
+		if err != nil {
+			message.SetError(err)
+			return nil, nil, message, nil
+		}
+
+		headerVal, err := extractHeaderValue(parsedEvent, ctx, property)
+		if err != nil {
+			message.SetError(err)
+			return nil, nil, message, nil
+		}
+		if headerVal != nil {
+			if message.HTTPHeaders == nil {
+				message.HTTPHeaders = make(map[string][]string)
+			}
+			message.HTTPHeaders[headerKey] = append(message.HTTPHeaders[headerKey], *headerVal)
+			return message, nil, nil, parsedEvent
+		}
+
+		return message, nil, nil, parsedEvent
+	}
+}
+
+func extractHeaderValue(parsedEvent analytics.ParsedEvent, ctx, prop string) (*string, error) {
+	spMap, err := parsedEvent.ToMap()
+	if err != nil {
+		return nil, err
+	}
+
+	gtmssPreview, ok := spMap[ctx]
```

Comment:
Could this be simplified by using the analytics SDK's [method for grabbing data from a context](https://github.com/snowplow/snowplow-golang-analytics-sdk/blob/a3430fbe576483d615b713120cfb5e443897d572/analytics/transform.go#L237)?

Reply:
Yes, i saw it also in use in the filter-context transformation. I chose to not use it because there is [an issue](https://github.com/snowplow/snowplow-golang-analytics-sdk/issues/36) reported for it, so i didn't want to count on it until that is fixed. Also `toMap` felt slightly neater/faster (no nested for loops and json marshalling), but i haven't tested this yet and i may be completely wrong.

Reply:
Chatted on a call - understand the rationale but it's probably better to just use the analytics SDK method for this one. The change related to that issue wouldn't be much of a problem, and I don't see any obvious efficiency benefit one way or the other (but we shouldn't spend time testing that)